### DISCUSSION TITLE
How to make the gradient field of the gravitaional potential or the pressure?

### DISCUSSION THREAD

**Discussion by applejwjcat** [osyris-project/osyris#116] [Q&A]
By using the osyris, I can easily to make some plots, but I want to obtain the gradient field of the gravitational potential or pressure, I didn't find the right way to do this. In my opinion, First I need to tansform the gravitational field from 1d to 3d,  then I choose one axis to get the gradient field. As I have known before, osyris avoid to make cube array as soon as possible, is there a convinent way using the osyris method to do this?

As well, I know the ```amrcube``` can give me the cube array, but I want to use python and the physical units, I only want to see the slice of  gradient field, I also want to save memery when I perform the ```1024``` data. 

Could you please give me some advice? Thanks a lot!

**Comment by nvaytet** [comment]
Hi there,
I guess it all depends on what you need the gradient for?
Do you need to compute the gradient at a few specific location in the 3D volume?
Do you need to make maps of the gradient?
Or do you need the gradient in a full 3D volume (around a given point?) to perform further computations?
Depending on the need, the solution will be more or less involved ;-)

**Reply by applejwjcat** [reply]
Hi,
I only want to make maps of the gradient, is there any solution? Please let me know, Thanks very much!

**Comment by nvaytet** [comment]
So I will assume the gradient you want to compute is 3D in nature, such as the one you would use to compute a hydrostatic equilibrium.

For making maps, I think the best would probably be to make something like a [thick map](https://osyris.readthedocs.io/en/stable/plotting_thick_maps.html) of e.g. pressure, and instead of flattening it, we would just return the values in the 3D (rectangular) volume. From that grid of pressure values, we could easily compute the gradient of pressure, for example using something like https://numpy.org/doc/stable/reference/generated/numpy.gradient.html

Right now there is no real way of extracting the cube of values, only the flattened values that were used to make the map (the default operation applied for a thick map is a sum along the third dimension) are returned. We would need an early return of `binned` [here](https://github.com/osyris-project/osyris/blob/main/src/osyris/plot/map.py#L339).

I am currently on easter vacation, and don't really have the opportunity to try this now. But it is making me think that we should probably have a set of lower level functions to compute all the values the maps need, and make those available through the API so that people can also use them directly, instead of having to go through the `map`. The `map` would then just use those functions internally.

One last question: did you want to make maps of the magnitude of the gradient (i.e. just an image of scalars) or do you need it as a vector and you wish to represent it as arrows or streamlines?

**Reply by nvaytet** [reply]
For the low-level functions, it could for example be a function that will compute a regular grid of points in a given plane or thick plane, and then a second function which would just interpolate in the AMR mesh the value at a set of given locations.
The `map` would just call those two functions one after the other.

**Comment by applejwjcat** [comment]
Thank you for your reply, I have tried to add some codes to your osyris program to extract the cube, but I am not familiar with the ```amr```, I will try again. otherwise I can use the ```amrcube``` to do the same things (not pythonism). I am very excited that you would give some low level API for us to use. 
Enjoy your vacation! Have fun !!!
